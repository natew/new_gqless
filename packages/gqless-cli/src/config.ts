import { cosmiconfig } from 'cosmiconfig';
import { promises } from 'fs';
import { resolve } from 'path';
import { Boolean, Partial, Static, String, Unknown } from 'runtypes';

import type { GenerateOptions } from './generate';
import type { IntrospectionOptions } from './introspection';

type GqlessCombinedOptions = GenerateOptions & {
  /**
   * Introspection options
   */
  introspection?: IntrospectionOptions;
  /**
   * Client generation destination
   */
  destination?: string;
};

export type GqlessConfig = Required<
  Static<typeof gqlessCLIConfigRecord>
> extends Required<GqlessCombinedOptions>
  ? GqlessCombinedOptions
  : never;

const StringRecord = Unknown.withConstraint<Record<string, string>>(
  (scalars) => {
    return (
      typeof scalars === 'object' &&
      scalars != null &&
      !Array.isArray(scalars) &&
      Object.values(scalars).every((v) => typeof v === 'string')
    );
  }
);

const gqlessCLIConfigRecord = Partial({
  endpoint: String,
  enumsAsStrings: Boolean,
  scalars: StringRecord,
  react: Boolean,
  preImport: String,
  introspection: Partial({
    endpoint: String,
    headers: StringRecord,
  }),
  destination: String,
  subscriptions: Boolean,
});

export const defaultConfig = {
  endpoint: '/api/graphql',
  enumsAsStrings: false,
  react: false,
  scalars: {
    DateTime: 'string',
  },
  preImport: '',
  introspection: {
    endpoint: 'SPECIFY_ENDPOINT_OR_SCHEMA_FILE_PATH_HERE',
    headers: {},
  } as IntrospectionOptions,
  destination: './src/generated/graphql.ts',
  subscriptions: false,
};

const defaultFilePath = resolve(process.cwd(), 'gqless.config.cjs');

type DeepReadonly<T> = T extends (infer R)[]
  ? DeepReadonlyArray<R>
  : T extends Function
  ? T
  : T extends object
  ? DeepReadonlyObject<T>
  : T;

interface DeepReadonlyArray<T> extends ReadonlyArray<DeepReadonly<T>> {}

type DeepReadonlyObject<T> = {
  readonly [P in keyof T]: DeepReadonly<T[P]>;
};

export const gqlessConfigPromise: Promise<{
  filepath: string;
  config: DeepReadonly<GqlessConfig>;
}> = cosmiconfig('gqless', {
  searchPlaces: ['gqless.config.cjs', 'gqless.config.js', 'package.json'],
})
  .search()
  .then(async (config) => {
    if (!config || config.isEmpty) {
      const filepath = config?.filepath || defaultFilePath;

      const NODE_ENV = process.env['NODE_ENV'];

      if (NODE_ENV !== 'test' && NODE_ENV !== 'production') {
        const { format } = (await import('./prettier')).formatPrettier({
          parser: 'typescript',
        });
        await promises.writeFile(
          defaultFilePath,
          await format(`
          /**
           * @type {import("@dish/gqless-cli").GqlessConfig}
           */
          const config = ${JSON.stringify(defaultConfig)};
          
          module.exports = config;`)
        );
      }
      return {
        filepath,
        config: defaultConfig,
      };
    }
    return {
      config: gqlessCLIConfigRecord.check(config.config),
      filepath: config.filepath,
    };
  })
  .catch(() => ({ config: defaultConfig, filepath: defaultFilePath }));
