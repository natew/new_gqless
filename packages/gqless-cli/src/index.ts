export * from "./generate";
export * from "./introspection";
export * from "./writeGenerate";
