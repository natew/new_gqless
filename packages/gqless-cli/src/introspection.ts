import { fetch } from 'cross-fetch';
import { print } from 'graphql';

import { AsyncExecutor } from '@graphql-tools/delegate';
import { introspectSchema, wrapSchema } from '@graphql-tools/wrap';

export const getRemoteSchema = async (
  /**
   * Endpoint of the remote GraphQL API, if not specified, it points to http://localhost:3000/graphql
   */
  endpoint: string,
  /**
   * Specify configuration like headers for the introspection
   */
  { headers = {} }: { headers?: Record<string, string> } = {}
) => {
  const executor: AsyncExecutor = async ({ document, variables }) => {
    const query = print(document);
    const fetchResult = await fetch(endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
      body: JSON.stringify({ query, variables }),
    });
    return fetchResult.json();
  };

  const schema = wrapSchema({
    schema: await introspectSchema(executor, {
      endpoint,
    }),
    executor,
  });

  return schema;
};
