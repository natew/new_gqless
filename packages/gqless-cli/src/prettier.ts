import {
  Options as PrettierOptions,
  resolveConfig,
  format as prettierFormat,
} from 'prettier';

const commonConfig = resolveConfig(process.cwd());

export function formatPrettier(
  defaultOptions: Omit<PrettierOptions, 'parser'> &
    Required<Pick<PrettierOptions, 'parser'>>
) {
  const configPromise = commonConfig.then((config) =>
    Object.assign({}, config, defaultOptions)
  );

  return {
    async format(input: string) {
      return prettierFormat(input, await configPromise);
    },
  };
}
