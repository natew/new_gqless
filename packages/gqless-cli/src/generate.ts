import {
  GraphQLEnumType,
  GraphQLInputObjectType,
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLSchema,
  parse,
  printSchema,
} from 'graphql';
import fromPairs from 'lodash/fromPairs';
import { format, Options as PrettierOptions, resolveConfig } from 'prettier';

import { parseSchemaType, ScalarsEnumsHash, Schema, Type } from '@dish/gqless';
import { codegen } from '@graphql-codegen/core';
import * as typescriptPlugin from '@graphql-codegen/typescript';

export type GenerateOptions = {
  /**
   * Overwrite the default 'queryFetcher'
   * 
   * @default
   * const queryFetcher: QueryFetcher = async function (query, variables) {
        const response = await fetch('/graphql', {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            query,
            variables,
          }),
          mode: 'cors',
        });

      if (!response.ok) {
        throw new Error(`Network error, received status code ${response.status}`);
      }

      const json = await response.json();

      return json;
    };
   */
  queryFetcher?: string;
  /**
   * Add a custom string at the beginning of the file, for example, add imports.
   */
  preImport?: string;
  /**
   * Specify the types of custom scalars
   */
  scalars?: Record<string, string>;
};

export async function generate(
  schema: GraphQLSchema,
  { queryFetcher, preImport = '', scalars }: GenerateOptions = {}
) {
  const prettierConfig = resolveConfig(process.cwd());
  const codegenResult = codegen({
    schema: parse(printSchema(schema)),
    config: {},
    documents: [],
    filename: 'gqless.generated.ts',
    pluginMap: {
      typescript: typescriptPlugin,
    },
    plugins: [
      {
        typescript: {
          onlyOperationTypes: true,
          declarationKind: 'interface',
          addUnderscoreToArgsType: true,
          scalars,
          namingConvention: 'keep',
        },
      },
    ],
  });

  const config = schema.toConfig();

  const scalarsEnumsHash: ScalarsEnumsHash = {};

  const enumsNames: string[] = [];

  const inputTypeNames = new Set<string>();

  const generatedSchema: Schema = {
    query: {},
    mutation: {},
    subscription: {},
  };

  const queryType = config.query;
  const mutationType = config.mutation;
  const subscriptionType = config.subscription;

  const parseEnumType = (type: GraphQLEnumType) => {
    scalarsEnumsHash[type.name] = true;
    enumsNames.push(type.name);
  };
  const parseScalarType = (type: GraphQLScalarType) => {
    scalarsEnumsHash[type.name] = true;
  };
  const parseObjectType = (type: GraphQLObjectType, typeName = type.name) => {
    const fields = type.getFields();

    const schemaType: Record<string, Type> = {};

    Object.entries(fields).forEach(([key, value]) => {
      schemaType[key] = {
        __type: value.type.toString(),
      };

      if (value.args.length) {
        schemaType[key].__args = fromPairs(
          value.args.map((arg) => {
            return [arg.name, arg.type.toString()];
          })
        );
      }
    });

    generatedSchema[typeName] = schemaType;
  };

  const parseInputType = (type: GraphQLInputObjectType) => {
    inputTypeNames.add(type.name);
    const fields = type.getFields();

    const schemaType: Record<string, Type> = {};

    Object.entries(fields).forEach(([key, value]) => {
      schemaType[key] = {
        __type: value.type.toString(),
      };
    });

    generatedSchema[type.name] = schemaType;
  };

  config.types.forEach((type) => {
    if (
      type.name.startsWith('__') ||
      type === queryType ||
      type === mutationType ||
      type === subscriptionType
    )
      return;

    if (type instanceof GraphQLEnumType) {
      parseEnumType(type);
    } else if (type instanceof GraphQLScalarType) {
      parseScalarType(type);
    } else if (type instanceof GraphQLObjectType) {
      parseObjectType(type);
    } else if (type instanceof GraphQLInputObjectType) {
      parseInputType(type);
    }
  });

  /* istanbul ignore else */
  if (queryType) {
    parseObjectType(queryType, 'query');
  }

  if (mutationType) {
    parseObjectType(mutationType, 'mutation');
  }

  if (subscriptionType) {
    parseObjectType(subscriptionType, 'subscription');
  }

  function parseFinalType({
    pureType,
    isArray,
    nullableItems,
    isNullable,
  }: ReturnType<typeof parseSchemaType>) {
    let typeToReturn: string[] = [
      scalarsEnumsHash[pureType] ? `ScalarsEnums["${pureType}"]` : pureType,
    ];

    if (isArray) {
      typeToReturn = [
        'Array<',
        ...(nullableItems ? ['Maybe<', ...typeToReturn, '>'] : typeToReturn),
        '>',
      ];
    }

    if (isNullable) {
      typeToReturn = ['Maybe<', ...typeToReturn, '>'];
    }

    return typeToReturn.join('');
  }

  let typescriptTypes = Object.entries(generatedSchema).reduce(
    (acum, [typeKey, typeValue]) => {
      const typeName = (() => {
        switch (typeKey) {
          case 'query': {
            return 'Query';
          }
          case 'mutation': {
            return 'Mutation';
          }
          case 'subscription': {
            return 'Subscription';
          }
          default: {
            return typeKey;
          }
        }
      })();

      if (inputTypeNames.has(typeName)) return acum;

      acum += `

      export interface ${typeName} {
        ${Object.entries(typeValue).reduce((acum, [fieldKey, fieldValue]) => {
          const fieldValueProps = parseSchemaType(fieldValue.__type);
          const typeToReturn = parseFinalType(fieldValueProps);
          if (fieldValue.__args) {
            const argsEntries = Object.entries(fieldValue.__args);
            let onlyNullableArgs = true;
            const argTypes = argsEntries.reduce(
              (acum, [argKey, argValue], index) => {
                const argValueProps = parseSchemaType(argValue);
                const connector = argValueProps.isNullable ? '?:' : ':';

                if (!argValueProps.isNullable) {
                  onlyNullableArgs = false;
                }

                const argTypeValue = parseFinalType(argValueProps);

                acum += `${argKey}${connector} ${argTypeValue}`;
                if (index < argsEntries.length - 1) {
                  acum += '; ';
                }
                return acum;
              },
              ''
            );
            const connector = onlyNullableArgs ? '?:' : ':';
            acum += `
            ${fieldKey}: (args${connector} {${argTypes}}) => ${typeToReturn}`;
          } else {
            acum += `
            ${fieldKey}: ${typeToReturn}`;
          }

          return acum;
        }, '')}
      }
      `;

      return acum;
    },
    ''
  );

  typescriptTypes += `
    export interface GeneratedSchema {
      query: Query
      mutation: Mutation
      subscription: Subscription
    }
    `;

  typescriptTypes += `
    export interface ScalarsEnums extends Scalars {
      ${enumsNames.reduce((acum, enumName) => {
        acum += `${enumName}: ${enumName};`;
        return acum;
      }, '')}
    }
    `;

  queryFetcher =
    queryFetcher ||
    `
    const queryFetcher : QueryFetcher = async function (query, variables) {
        const response = await fetch("/graphql", {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            query,
            variables,
          }),
          mode: "cors",
        });
      
        if (!response.ok) {
          throw new Error(\`Network error, received status code \${response.status}\`);
        }
      
        const json = await response.json();
      
        return json;
      };
    `;

  const code = format(
    `
  ${preImport}
  import { createClient, QueryFetcher, ScalarsEnumsHash } from "@dish/gqless";

  ${await codegenResult}

  const scalarsEnumsHash: ScalarsEnumsHash = ${JSON.stringify(
    scalarsEnumsHash
  )};
  export const generatedSchema = ${JSON.stringify(generatedSchema)} as const;

  ${typescriptTypes}

  ${queryFetcher}

  export const client = createClient<GeneratedSchema>(generatedSchema, scalarsEnumsHash, queryFetcher)

  export const { query, mutation, subscription, resolved, selectFields } = client;
  `,
    Object.assign({}, await prettierConfig, {
      parser: 'typescript',
    } as PrettierOptions)
  );

  return {
    code,
    generatedSchema,
    scalarsEnumsHash,
  };
}
