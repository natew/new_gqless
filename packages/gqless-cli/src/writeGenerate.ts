import fs from 'fs';
import { GraphQLSchema } from 'graphql';
import mkdirp from 'mkdirp';
import { dirname, resolve } from 'path';

import { generate, GenerateOptions } from './generate';

async function writeClientCode({
  destinationPath,
  clientCode,
}: {
  clientCode: string;
  destinationPath: string;
}): Promise<void> {
  if (fs.existsSync(destinationPath)) return;

  await fs.promises.writeFile(destinationPath, clientCode, {
    encoding: 'utf-8',
  });
}

async function writeSchemaCode({
  schemaCode,
  destinationPath,
}: {
  schemaCode: string;
  destinationPath: string;
}): Promise<void> {
  const schemaPath = resolve(dirname(destinationPath), './schema.generated.ts');

  if (fs.existsSync(schemaPath)) {
    const existingCode = await fs.promises.readFile(schemaPath, {
      encoding: 'utf-8',
    });

    if (existingCode === schemaCode) return;
  }

  await fs.promises.writeFile(schemaPath, schemaCode, {
    encoding: 'utf-8',
  });
}

export async function writeGenerate(
  schema: GraphQLSchema,
  destinationPath: string,
  generateOptions?: GenerateOptions
) {
  if (!destinationPath.endsWith('.ts')) {
    const err = Error(
      'You have to specify the ".ts" extension, instead, it received: "' +
        destinationPath +
        '"'
    );

    Error.captureStackTrace(err, writeGenerate);

    throw err;
  }

  destinationPath = resolve(destinationPath);

  const [{ clientCode, schemaCode }] = await Promise.all([
    generate(schema, generateOptions),
    mkdirp(dirname(destinationPath)),
  ]);

  await Promise.all([
    writeClientCode({ clientCode, destinationPath }),
    writeSchemaCode({ schemaCode, destinationPath }),
  ]);

  return destinationPath;
}
