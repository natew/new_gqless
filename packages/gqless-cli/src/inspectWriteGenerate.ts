import { existsSync } from 'fs';
import { resolve } from 'path';

import { GenerateOptions } from './generate';
import { getRemoteSchema } from './introspection';
import { writeGenerate } from './writeGenerate';

export async function inspectWriteGenerate({
  endpoint,
  overwrite,
  destination = './src/generated/index.ts',
  generateOptions,
  cli,
  headers,
}: {
  /**
   * GraphQL API endpoint
   *
   * @example 'http://localhost:3000/graphql'
   */
  endpoint: string;
  /**
   * Whether the generation should overwrite if file already exists
   */
  overwrite?: boolean;
  /**
   * File path destination
   * @default './src/generated/index.ts'
   */
  destination?: string;
  /**
   * Specify generate options
   */
  generateOptions?: GenerateOptions;
  /**
   * Whether it's being called through the CLI
   */
  cli?: boolean;
  /**
   * Specify headers for the introspection HTTP request
   */
  headers?: Record<string, string>;
}) {
  destination = resolve(destination);

  const schema = await getRemoteSchema(endpoint, {
    headers,
  });

  if (!overwrite && existsSync(destination)) {
    const err = Error(
      `File '${destination}' already exists, specify ${
        cli ? "'--overwrite'" : "'overwrite: true'"
      } to overwrite the existing file.`
    );

    Error.captureStackTrace(err, inspectWriteGenerate);

    throw err;
  }

  const generatedPath = await writeGenerate(
    schema,
    destination,
    generateOptions
  );

  if (cli) {
    console.log('Code generated successfully at ' + generatedPath);
  }
}
