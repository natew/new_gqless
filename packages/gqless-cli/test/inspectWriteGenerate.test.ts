import fs from 'fs';
import { createTestApp } from 'test-utils';

import { inspectWriteGenerate } from '../src/inspectWriteGenerate';
import { getTempDir } from './utils';

const { readFile, writeFile } = fs.promises;
const { server, isReady } = createTestApp({
  schema: `
    type Query {
        hello: String!
    }
    `,
  resolvers: {
    Query: {
      hello() {
        return 'hello world';
      },
    },
  },
});

let endpoint: string;
beforeAll(async () => {
  await isReady;

  endpoint = (await server.listen(0)) + '/graphql';
});

afterAll(async () => {
  server.close();
});

test('basic inspectWriteGenerate functionality', async () => {
  const tempDir = await getTempDir();

  try {
    await inspectWriteGenerate({
      endpoint,
      overwrite: true,
      destination: tempDir.clientPath,
    });

    expect(
      await readFile(tempDir.clientPath, {
        encoding: 'utf-8',
      })
    ).toMatchSnapshot('basic inspectWriteGenerate client');

    expect(
      await readFile(tempDir.schemaPath, {
        encoding: 'utf-8',
      })
    ).toMatchSnapshot('basic inspectWriteGenerate schema');
  } finally {
    await tempDir.cleanup();
  }
});

test('specify generateOptions to inspectWriteGenerate', async () => {
  const tempDir = await getTempDir();

  const shouldBeIncluded = '// This should be included';

  try {
    await inspectWriteGenerate({
      endpoint,
      overwrite: true,
      destination: tempDir.clientPath,
      generateOptions: {
        preImport: `
            ${shouldBeIncluded}
            `,
      },
    });

    const generatedFileContentClient = await readFile(tempDir.clientPath, {
      encoding: 'utf-8',
    });

    const generatedFileContentSchema = await readFile(tempDir.schemaPath, {
      encoding: 'utf-8',
    });

    expect(generatedFileContentClient).toMatchSnapshot(
      'generateOptions client'
    );
    expect(generatedFileContentSchema).toMatchSnapshot(
      'generateOptions schema'
    );

    expect(
      generatedFileContentSchema.startsWith(shouldBeIncluded)
    ).toBeTruthy();
  } finally {
    await tempDir.cleanup();
  }
});

describe('inspect headers', () => {
  let endpoint: string;

  const secretToken = 'super secret token';

  const { server, isReady } = createTestApp({
    schema: `
        type Query {
            hello: String!
        }
        `,
    resolvers: {
      Query: {
        hello() {
          return 'hello world';
        },
      },
    },
    context: (req, _reply) => {
      if (req.headers.authorization !== secretToken) {
        throw Error('Unauthorized!');
      }
      return {};
    },
  });

  beforeAll(async () => {
    await isReady;
    endpoint = (await server.listen(0)) + '/graphql';
  });

  afterAll(async () => {
    await server.close();
  });

  test('specify headers to inspectWriteGenerate', async () => {
    const tempDir = await getTempDir();

    const shouldBeIncluded = '// This should be included';

    try {
      await inspectWriteGenerate({
        endpoint,
        overwrite: true,
        destination: tempDir.clientPath,
        headers: {
          authorization: secretToken,
        },
        generateOptions: {
          preImport: shouldBeIncluded,
        },
      });

      const generatedFileContent = await readFile(tempDir.schemaPath, {
        encoding: 'utf-8',
      });

      expect(generatedFileContent).toMatchSnapshot('specify headers');

      expect(generatedFileContent.startsWith(shouldBeIncluded)).toBeTruthy();
    } finally {
      await tempDir.cleanup();
    }
  });

  test('should throw if headers are not specified when required by server', async () => {
    const tempDir = await getTempDir({
      initClientFile: '',
    });

    try {
      await expect(
        inspectWriteGenerate({
          endpoint,
          overwrite: true,
          destination: tempDir.clientPath,
        })
      ).rejects.toEqual({
        message: 'Unauthorized!',
      });

      const generatedFileContent = await readFile(tempDir.clientPath, {
        encoding: 'utf-8',
      });

      expect(generatedFileContent).toBe('');
    } finally {
      await tempDir.cleanup();
    }
  });
});

test('should respect to not overwrite', async () => {
  const tempDir = await getTempDir();

  try {
    await writeFile(tempDir.clientPath, 'this should be kept', {
      encoding: 'utf-8',
    });

    await expect(
      inspectWriteGenerate({
        endpoint,
        overwrite: false,
        destination: tempDir.clientPath,
      })
    ).rejects.toThrow(
      `File '${tempDir.clientPath}' already exists, specify 'overwrite: true' to overwrite the existing file.`
    );

    expect(
      await readFile(tempDir.clientPath, {
        encoding: 'utf-8',
      })
    ).toBe('this should be kept');
  } finally {
    await tempDir.cleanup();
  }
});

describe('CLI behavior', () => {
  test('overwrite message', async () => {
    const tempDir = await getTempDir();

    try {
      await writeFile(tempDir.clientPath, 'this should be kept', {
        encoding: 'utf-8',
      });

      await expect(
        inspectWriteGenerate({
          endpoint,
          overwrite: false,
          destination: tempDir.clientPath,
          cli: true,
        })
      ).rejects.toThrow(
        `File '${tempDir.clientPath}' already exists, specify '--overwrite' to overwrite the existing file.`
      );

      expect(
        await readFile(tempDir.clientPath, {
          encoding: 'utf-8',
        })
      ).toBe('this should be kept');
    } finally {
      await tempDir.cleanup();
    }
  });

  test('final message', async () => {
    const spy = jest.spyOn(console, 'log').mockImplementation();
    const tempDir = await getTempDir();

    try {
      await inspectWriteGenerate({
        endpoint,
        overwrite: true,
        destination: tempDir.clientPath,
        cli: true,
      });

      expect(spy).toHaveBeenCalledTimes(1);

      expect(spy).toHaveBeenLastCalledWith(
        'Code generated successfully at ' + tempDir.clientPath
      );

      expect(
        await readFile(tempDir.clientPath, {
          encoding: 'utf-8',
        })
      ).toMatchSnapshot('basic functionality with cli final messsage');

      expect(
        await readFile(tempDir.schemaPath, {
          encoding: 'utf-8',
        })
      ).toMatchSnapshot('basic functionality with cli final messsage - schema');
    } finally {
      await tempDir.cleanup();
      spy.mockRestore();
    }
  });
});
