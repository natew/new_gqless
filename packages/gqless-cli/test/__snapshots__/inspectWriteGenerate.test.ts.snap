// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI behavior final message: basic functionality with cli final messsage 1`] = `
"import { createClient, QueryFetcher, ScalarsEnumsHash } from '@dish/gqless';

export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export interface Scalars {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
}

export const scalarsEnumsHash: ScalarsEnumsHash = {
  String: true,
  Boolean: true,
};
export const generatedSchema = {
  query: { __typename: { __type: 'String!' }, hello: { __type: 'String!' } },
  mutation: {},
  subscription: {},
} as const;

export interface Query {
  __typename: 'Query';
  hello: ScalarsEnums['String'];
}

export interface Mutation {
  __typename: 'Mutation';
}

export interface Subscription {
  __typename: 'Subscription';
}

export interface GeneratedSchema {
  query: Query;
  mutation: Mutation;
  subscription: Subscription;
}

export interface ScalarsEnums extends Scalars {}

const queryFetcher: QueryFetcher = async function (query, variables) {
  const response = await fetch('/graphql', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query,
      variables,
    }),
    mode: 'cors',
  });

  if (!response.ok) {
    throw new Error(\`Network error, received status code \${response.status}\`);
  }

  const json = await response.json();

  return json;
};

export const client = createClient<GeneratedSchema>({
  schema: generatedSchema,
  scalarsEnumsHash,
  queryFetcher,
});

export const { query, mutation, subscription, resolved, refetch } = client;
"
`;

exports[`basic inspectWriteGenerate functionality: basic inspectWriteGenerate 1`] = `
"import { createClient, QueryFetcher, ScalarsEnumsHash } from '@dish/gqless';

export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export interface Scalars {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
}

export const scalarsEnumsHash: ScalarsEnumsHash = {
  String: true,
  Boolean: true,
};
export const generatedSchema = {
  query: { __typename: { __type: 'String!' }, hello: { __type: 'String!' } },
  mutation: {},
  subscription: {},
} as const;

export interface Query {
  __typename: 'Query';
  hello: ScalarsEnums['String'];
}

export interface Mutation {
  __typename: 'Mutation';
}

export interface Subscription {
  __typename: 'Subscription';
}

export interface GeneratedSchema {
  query: Query;
  mutation: Mutation;
  subscription: Subscription;
}

export interface ScalarsEnums extends Scalars {}

const queryFetcher: QueryFetcher = async function (query, variables) {
  const response = await fetch('/graphql', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query,
      variables,
    }),
    mode: 'cors',
  });

  if (!response.ok) {
    throw new Error(\`Network error, received status code \${response.status}\`);
  }

  const json = await response.json();

  return json;
};

export const client = createClient<GeneratedSchema>({
  schema: generatedSchema,
  scalarsEnumsHash,
  queryFetcher,
});

export const { query, mutation, subscription, resolved, refetch } = client;
"
`;

exports[`inspect headers specify headers to inspectWriteGenerate: specify headers 1`] = `
"// This should be included
import { createClient, QueryFetcher, ScalarsEnumsHash } from '@dish/gqless';

export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export interface Scalars {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
}

export const scalarsEnumsHash: ScalarsEnumsHash = {
  String: true,
  Boolean: true,
};
export const generatedSchema = {
  query: { __typename: { __type: 'String!' }, hello: { __type: 'String!' } },
  mutation: {},
  subscription: {},
} as const;

export interface Query {
  __typename: 'Query';
  hello: ScalarsEnums['String'];
}

export interface Mutation {
  __typename: 'Mutation';
}

export interface Subscription {
  __typename: 'Subscription';
}

export interface GeneratedSchema {
  query: Query;
  mutation: Mutation;
  subscription: Subscription;
}

export interface ScalarsEnums extends Scalars {}

const queryFetcher: QueryFetcher = async function (query, variables) {
  const response = await fetch('/graphql', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query,
      variables,
    }),
    mode: 'cors',
  });

  if (!response.ok) {
    throw new Error(\`Network error, received status code \${response.status}\`);
  }

  const json = await response.json();

  return json;
};

export const client = createClient<GeneratedSchema>({
  schema: generatedSchema,
  scalarsEnumsHash,
  queryFetcher,
});

export const { query, mutation, subscription, resolved, refetch } = client;
"
`;

exports[`specify generateOptions to inspectWriteGenerate: generateOptions 1`] = `
"// This should be included

import { createClient, QueryFetcher, ScalarsEnumsHash } from '@dish/gqless';

export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export interface Scalars {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
}

export const scalarsEnumsHash: ScalarsEnumsHash = {
  String: true,
  Boolean: true,
};
export const generatedSchema = {
  query: { __typename: { __type: 'String!' }, hello: { __type: 'String!' } },
  mutation: {},
  subscription: {},
} as const;

export interface Query {
  __typename: 'Query';
  hello: ScalarsEnums['String'];
}

export interface Mutation {
  __typename: 'Mutation';
}

export interface Subscription {
  __typename: 'Subscription';
}

export interface GeneratedSchema {
  query: Query;
  mutation: Mutation;
  subscription: Subscription;
}

export interface ScalarsEnums extends Scalars {}

const queryFetcher: QueryFetcher = async function (query, variables) {
  const response = await fetch('/graphql', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query,
      variables,
    }),
    mode: 'cors',
  });

  if (!response.ok) {
    throw new Error(\`Network error, received status code \${response.status}\`);
  }

  const json = await response.json();

  return json;
};

export const client = createClient<GeneratedSchema>({
  schema: generatedSchema,
  scalarsEnumsHash,
  queryFetcher,
});

export const { query, mutation, subscription, resolved, refetch } = client;
"
`;
